docs/adr/010_device_request_queue.md

```diff
@@ -0,0 +1,149 @@
+- There are no operations starting around the same time that would cause it to
+  be opened again.
+
+The following cases can trigger operations from different threads:
+
+- An OTAA join message; `DeviceJoinLoader` eventually loads the twins and
+  stores them back up.
+
+- A data message; `DeviceLoaderSynchronizer` loads the twins for multiple
+  devices based on the `DevAddr` and adds them to the cache.
+
+- C2D message that needs to be delivered can trigger a load if the device is
+  not in the cache.
+
+- Cache refreshes can happen on a background thread for a device, causing the
+  need for a connection.
+
+There are a number of complex and concurrent code paths in the current
+solution that make it hard to reason about when a device connection is in use,
+when it's safe to open/close without affecting another operation in flight and
+how various device-related operations (such as refreshing from the twin) can
+work in isolation, deterministically and without race conditions. If a single
+queue could be maintained for all device-related operations then it would
+become easier to order those operations, reason about them and make they don't
+cause connections to be opened when an LNS has lost the race against another.
+
+Several approaches were explored to understand the overall impact of
+refactoring, whether the changes to the current code base and design would be
+large or small and worth the benefits they bring. The approaches explored can
+be summed up as follows:
+
+1. Add a queue to `LoRaDevice` for all device-related operations. There is one
+   that exists today but it is used to service uplink data requests only. The
+   idea would be to extend it to encompass all other requests and operations.
+
+1. Turn the LNS implementation to be (logically) single-threaded so that not
+   only does it make it easier to reason about the code (except perhaps with
+   regards to re-entrancy), but it also enables use of very simple data
+   structures without the need for locking. This is possible because the bulk
+   of LNS is I/O bound. There is practically no CPU-intensive code that
+   executes between each await.
+
+1. Simplify the entire flow and processing of a message into a simple
+   request-response model that can be manipulated and reasoned about much more
+   easily. That is, all methods receiving and processing a `LoRaRequest`
+   return `Task<DownlinkMessage>` instead of `void` or `Task`. This allows all
+   `Task.Run` uses to be moved to a single and central point when a message is
+   received instead of being littered throughout the code base. The main
+   message loop could also be made responsible for central
+   error-handling/logging and sending of downlink messages when the processing
+   of a request has completed. This would just require regular use of tasks.
+   Next, all `Task`-based operations on `LoRaDevice` could be naturally and
+   _implicitly_ queued and then executed in a mutually exclusive manner.
+
+## Decision
```

Comment:
the decision needs to be updated with what we discussed yesterday.

Reply:
Done with b70734e222bc36fe66f22d6da59c7bfc8dee74dd.